### PR TITLE
Lua API: Add sound_pause and sound_resume to minetest api

### PR DESCRIPTION
**Pull request by 14NGiestas**
Hello folks!
I'm learning C++ so I decided to practice adding this to minetest lua api... My first issue in this repo was about adding a range option in minetest.sound_play, however I realized (after all this time) that a range option isn't needed to solve my initial problem.
I'm so happy I did something barely useful in C++ in a huge project... after many tries, energy drinks and a while reading the docs of OpenAL and Minetest core stuff... searching in a bunch of files :sleepy:
Well... whatever, let's show the stuff:

Exemple use (dunno why I used a "c" table):
```
c = {}
c.id = 0

minetest.register_chatcommand("play", {
	params = "<text>",
	description = "playsound",
	func = function( _ , text)
		c.id = minetest.sound_play("teste_q") -- Nothing new here uh?
		print(c.id)
	end,
})

minetest.register_chatcommand("pause", {
	description = "pause sound",
	func = function( _ )
		minetest.sound_pause(c.id) -- This pause the sound
	end,
})

minetest.register_chatcommand("resume", {
	description = "resume sound",
	func = function( _ )
		minetest.sound_resume(c.id) -- This resume the paused sound
	end,
})

minetest.register_chatcommand("stop", {
	description = "stop sound",
	func = function( _ )
		minetest.sound_stop(c.id) -- This completely stop the sound
	end,
})
```

### CODE REVIEW

File: src/sound_openal.cpp [hunk 1]
```diff
@@ -454,6 +454,23 @@ class OpenALSoundManager: public ISoundManager
...
 
+	void pauseSnd(int id)
+	{
+		UNORDERED_MAP<int, PlayingSound*>::iterator i = m_sounds_playing.find(id);
+		if(i == m_sounds_playing.end())
+			return;
+		PlayingSound *sound = i->second;
+		alSourcePause(sound->source_id);
+	}
+	void resumeSnd(int id)
```
Comment (nerzhul) [inline src/sound_openal.cpp]: missing empty line

File: src/sound_openal.cpp [hunk 2]
```diff
@@ -454,6 +454,23 @@ class OpenALSoundManager: public ISoundManager
...
+	{
+		UNORDERED_MAP<int, PlayingSound*>::iterator i = m_sounds_playing.find(id);
+		if(i == m_sounds_playing.end())
```
Comment (nerzhul) [inline src/sound_openal.cpp]: missing space

File: src/sound_openal.cpp [hunk 3]
```diff
@@ -573,6 +592,16 @@ class OpenALSoundManager: public ISoundManager
 		maintain();
 		deleteSound(sound);
 	}
+	void pauseSound(int sound)
+	{
+		maintain();
+		pauseSnd(sound);
+	}
+	void resumeSound(int sound)
```
Comment (nerzhul) [inline src/sound_openal.cpp]: missing empty line

File: src/sound_openal.cpp [hunk 4]
```diff
@@ -484,15 +501,17 @@ class OpenALSoundManager: public ISoundManager
...
 			int id = i->first;
 			PlayingSound *sound = i->second;
 			// If not playing, remove it
 			{
 				ALint state;
 				alGetSourcei(sound->source_id, AL_SOURCE_STATE, &state);
-				if(state != AL_PLAYING){
-					del_list.insert(id);
+				if(state != AL_PLAYING && state != AL_PAUSED)
+                {
```
Comment (nerzhul) [inline src/sound_openal.cpp]: code style problem, brace + space

File: src/network/clientopcodes.cpp [hunk 5]
```diff
@@ -108,8 +108,8 @@ const ToClientCommandHandler toClientCommandTable[TOCLIENT_NUM_MSG_TYPES] =
 	{ "TOCLIENT_LOCAL_PLAYER_ANIMATIONS",  TOCLIENT_STATE_CONNECTED, &Client::handleCommand_LocalPlayerAnimations }, // 0x51
 	{ "TOCLIENT_EYE_OFFSET",               TOCLIENT_STATE_CONNECTED, &Client::handleCommand_EyeOffset }, // 0x52
 	{ "TOCLIENT_DELETE_PARTICLESPAWNER",   TOCLIENT_STATE_CONNECTED, &Client::handleCommand_DeleteParticleSpawner }, // 0x53
-	null_command_handler,
-	null_command_handler,
+	{ "TOCLIENT_PAUSE_SOUND",              TOCLIENT_STATE_CONNECTED, &Client::handleCommand_PauseSound }, // 0x54
```
Comment (nerzhul) [inline src/network/clientopcodes.cpp]: you forgot sending rule into serveropcodes.cpp

File: src/client.h [hunk 6]
```diff
@@ -388,6 +388,8 @@ class Client : public con::PeerHandler, public InventoryManager, public IGameDef
 	void handleCommand_ItemDef(NetworkPacket* pkt);
 	void handleCommand_PlaySound(NetworkPacket* pkt);
 	void handleCommand_StopSound(NetworkPacket* pkt);
+	void handleCommand_PauseSound(NetworkPacket* pkt);
+    void handleCommand_ResumeSound(NetworkPacket* pkt);
```
Comment (nerzhul) [inline src/client.h]: wrong indent

File: src/script/lua_api/l_server.cpp [hunk 7]
```diff
@@ -429,6 +429,7 @@ int ModApiServer::l_sound_play(lua_State *L)
 	return 1;
 }
 
+
```
Comment (nerzhul) [inline src/script/lua_api/l_server.cpp]: useless

**Comment by paramat** [comment]
What are the uses for these?

**Comment by rubenwardy** [comment]
Stopping long or looping music, pretty much.
Ie: juke boxes

File: src/network/clientopcodes.cpp (hunk 5 above)
Comment (14NGiestas) [inline src/network/clientopcodes.cpp]: Done!

File: src/sound_openal.cpp [hunk 8]
```diff
@@ -491,7 +511,8 @@ class OpenALSoundManager: public ISoundManager
 			{
 				ALint state;
 				alGetSourcei(sound->source_id, AL_SOURCE_STATE, &state);
-				if(state != AL_PLAYING){
+				if(state != AL_PLAYING && state != AL_PAUSED)
+				{
```
Comment (nerzhul) [inline src/sound_openal.cpp]: brace on wrong line

File: src/sound_openal.cpp [hunk 9]
```diff
@@ -511,8 +511,7 @@ class OpenALSoundManager: public ISoundManager
 			{
 				ALint state;
 				alGetSourcei(sound->source_id, AL_SOURCE_STATE, &state);
-				if(state != AL_PLAYING && state != AL_PAUSED)
-				{
+				if(state != AL_PLAYING && state != AL_PAUSED) {
```
Comment (nerzhul) [inline src/sound_openal.cpp]: missing space after if, please read our code style guidelines

File: src/network/clientpackethandler.cpp [hunk 10]
```diff
@@ -819,6 +819,36 @@ void Client::handleCommand_StopSound(NetworkPacket* pkt)
...
 
+
+
+void Client::handleCommand_PauseSound(NetworkPacket* pkt)
+{
+	s32 server_id;
+
+	*pkt >> server_id;
+
+	UNORDERED_MAP<s32, int>::iterator i = m_sounds_server_to_client.find(server_id);
```
Comment (rubenwardy) [inline src/network/clientpackethandler.cpp]: please use `it` as it's an iterator not a counter (`i`)

File: src/sound_openal.cpp (hunk 9 above)
Comment (14NGiestas) [inline src/sound_openal.cpp]: Oh sorry about that @nerzhul I'm really new here.
I'm reading the code guidelines now (What a shame).
I hope there's no more heresies left.

File: src/network/clientpackethandler.cpp (hunk 10 above)
Comment (14NGiestas) [inline src/network/clientpackethandler.cpp]: @rubenwardy this one is the same as every original iterator name in the minetest sound api stuff.
If this is inaccurate, should I change every iterator I've seen labeled as `i` (like in some functions above this one)?

**Comment by 14NGiestas** [comment]
Any issue left?

**Comment by MT-Modder** [comment]
Jukeboxes would work much better if you could get the length of the sound file...

File: src/sound_openal.cpp [hunk 11]
```diff
@@ -454,6 +454,26 @@ class OpenALSoundManager: public ISoundManager
 		m_sounds_playing.erase(id);
 	}
 
+	void pauseSnd(int id)
+	{
+		UNORDERED_MAP<int, PlayingSound*>::iterator i = m_sounds_playing.find(id);
+		if(i == m_sounds_playing.end())
```
Comment (rubenwardy) [inline src/sound_openal.cpp]: `if (`

File: src/sound_openal.cpp [hunk 12]
```diff
@@ -454,6 +454,26 @@ class OpenALSoundManager: public ISoundManager
 		m_sounds_playing.erase(id);
 	}
 
+	void pauseSnd(int id)
+	{
+		UNORDERED_MAP<int, PlayingSound*>::iterator i = m_sounds_playing.find(id);
```
Comment (rubenwardy) [inline src/sound_openal.cpp]: `it`

File: src/sound_openal.cpp [hunk 13]
```diff
@@ -454,6 +454,26 @@ class OpenALSoundManager: public ISoundManager
...
+		if(i == m_sounds_playing.end())
+			return;
+		PlayingSound *sound = i->second;
+		
+		alSourcePause(sound->source_id);
+	}
+	
+	void resumeSnd(int id)
+	{
+		UNORDERED_MAP<int, PlayingSound*>::iterator i = m_sounds_playing.find(id);
```
Comment (rubenwardy) [inline src/sound_openal.cpp]: `it`

File: src/sound_openal.cpp [hunk 14]
```diff
@@ -491,7 +511,8 @@ class OpenALSoundManager: public ISoundManager
 			{
 				ALint state;
 				alGetSourcei(sound->source_id, AL_SOURCE_STATE, &state);
-				if(state != AL_PLAYING){
+				
+				if(state != AL_PLAYING && state != AL_PAUSED) {
```
Comment (rubenwardy) [inline src/sound_openal.cpp]: `if (`

File: src/client.h [hunk 15]
```diff
@@ -331,7 +331,9 @@ class Client : public con::PeerHandler, public InventoryManager, public IGameDef
 	void handleCommand_ItemDef(NetworkPacket* pkt);
 	void handleCommand_PlaySound(NetworkPacket* pkt);
 	void handleCommand_StopSound(NetworkPacket* pkt);
-	void handleCommand_FadeSound(NetworkPacket *pkt);
+	void handleCommand_PauseSound(NetworkPacket* pkt);
```
Comment (Dumbeldor) [inline src/client.h]: void handleCommand_PauseSound(NetworkPacket *pkt);

File: src/client.h [hunk 16]
```diff
@@ -331,7 +331,9 @@ class Client : public con::PeerHandler, public InventoryManager, public IGameDef
...
+	void handleCommand_ResumeSound(NetworkPacket* pkt);
```
Comment (Dumbeldor) [inline src/client.h]: void handleCommand_PauseSound(NetworkPacket *pkt);

File: src/network/clientpackethandler.cpp [hunk 17]
```diff
@@ -821,13 +821,40 @@ void Client::handleCommand_StopSound(NetworkPacket* pkt)
 
 	*pkt >> server_id;
 
-	std::unordered_map<s32, int>::iterator i = m_sounds_server_to_client.find(server_id);
-	if (i != m_sounds_server_to_client.end()) {
-		int client_id = i->second;
+	std::unordered_map<s32, int>::iterator it = m_sounds_server_to_client.find(server_id);
```
Comment (Dumbeldor) [inline src/network/clientpackethandler.cpp]: const std::unordered_map

File: src/network/clientpackethandler.cpp [hunk 18]
```diff
@@ -821,13 +821,40 @@ void Client::handleCommand_StopSound(NetworkPacket* pkt)
...
+	}
+}
+
+void Client::handleCommand_ResumeSound(NetworkPacket* pkt)
+{
+	s32 server_id;
+
+	*pkt >> server_id;
+
+	std::unordered_map<s32, int>::iterator it = m_sounds_server_to_client.find(server_id);
```
Comment (Dumbeldor) [inline src/network/clientpackethandler.cpp]: const std::unordered_map

File: src/sound_openal.cpp [hunk 19]
```diff
@@ -159,7 +159,7 @@ SoundBuffer *load_opened_ogg_file(OggVorbis_File *oggFile,
 
 	ALenum error = alGetError();
 
-	if(error != AL_NO_ERROR){
+	if (error != AL_NO_ERROR){
```
Comment (Dumbeldor) [inline src/sound_openal.cpp]: add space before {

File: src/sound_openal.cpp [hunk 20]
```diff
@@ -302,14 +302,14 @@ class OpenALSoundManager: public ISoundManager
 		infostream<<"Audio: Initializing..."<<std::endl;
 
 		m_device = alcOpenDevice(NULL);
-		if(!m_device){
+		if (!m_device){
```
Comment (Dumbeldor) [inline src/sound_openal.cpp]: add space before {

File: src/sound_openal.cpp [hunk 21]
```diff
@@ -302,14 +302,14 @@ class OpenALSoundManager: public ISoundManager
...
 			infostream<<"Audio: No audio device available, audio system "
 				<<"not initialized"<<std::endl;
 			return;
 		}
 
 		m_context = alcCreateContext(m_device, NULL);
-		if(!m_context){
+		if (!m_context){
```
Comment (Dumbeldor) [inline src/sound_openal.cpp]: add space before {

File: src/sound_openal.cpp [hunk 22]
```diff
@@ -463,35 +465,55 @@ class OpenALSoundManager: public ISoundManager
 
 	void deleteSound(int id)
 	{
-		std::unordered_map<int, PlayingSound*>::iterator i = m_sounds_playing.find(id);
-		if(i == m_sounds_playing.end())
+		std::unordered_map<int, PlayingSound*>::iterator it = m_sounds_playing.find(id);
```
Comment (Dumbeldor) [inline src/sound_openal.cpp]: const std::unordered_map
and
PlayingSound *>

File: src/sound_openal.cpp [hunk 23]
```diff
@@ -463,35 +465,55 @@ class OpenALSoundManager: public ISoundManager
...
 
 		alDeleteSources(1, &sound->source_id);
 
 		delete sound;
 		m_sounds_playing.erase(id);
 	}
 
+	void pauseSnd(int id)
+	{
+		std::unordered_map<int, PlayingSound*>::iterator it = m_sounds_playing.find(id);
```
Comment (Dumbeldor) [inline src/sound_openal.cpp]: const std::unordered_map
and
PlayingSound *>

File: src/sound_openal.cpp [hunk 24]
```diff
@@ -384,11 +384,11 @@ class OpenALSoundManager: public ISoundManager
 
 	SoundBuffer* getBuffer(const std::string &name)
 	{
-		std::unordered_map<std::string, std::vector<SoundBuffer*>>::iterator i =
+		std::unordered_map<std::string, std::vector<SoundBuffer*> >::iterator it =
```
Comment (Dumbeldor) [inline src/sound_openal.cpp]: SoundBuffer *>

File: src/sound_openal.cpp [hunk 25]
```diff
@@ -384,11 +384,11 @@ class OpenALSoundManager: public ISoundManager
 
 	SoundBuffer* getBuffer(const std::string &name)
```
Comment (Dumbeldor) [inline src/sound_openal.cpp]: SoundBuffer *getBuffer

File: src/sound_openal.cpp [hunk 26]
```diff
@@ -370,10 +370,10 @@ class OpenALSoundManager: public ISoundManager
 
 	void addBuffer(const std::string &name, SoundBuffer *buf)
 	{
-		std::unordered_map<std::string, std::vector<SoundBuffer*>>::iterator i =
+		std::unordered_map<std::string, std::vector<SoundBuffer*> >::iterator it =
```
Comment (Dumbeldor) [inline src/sound_openal.cpp]: <SoundBuffer *>

File: src/sound_openal.cpp [hunk 27]
```diff
@@ -645,13 +681,14 @@ class OpenALSoundManager: public ISoundManager
 		maintain();
 		return (m_sounds_playing.count(sound) != 0);
 	}
-
+	
 	void updateSoundPosition(int id, v3f pos)
 	{
-		std::unordered_map<int, PlayingSound*>::iterator i = m_sounds_playing.find(id);
-		if (i == m_sounds_playing.end())
+		std::unordered_map<int, PlayingSound*>::iterator it = m_sounds_playing.find(id);
```
Comment (Dumbeldor) [inline src/sound_openal.cpp]: const std::unordered_map

**Comment by rubenwardy** [comment]
Still interested in working on this?

**Comment by paramat** [comment]
Note we can close PRs if no author response in a month.
@14NGiestas

**Comment by 14NGiestas** [comment]
Can I reopen another time? I'm a little bit busy (also lost my local
files). But I still interested in those features.

What you all think?

Em 04/12/2017 02:19, "Paramat" <notifications@github.com> escreveu:

Note we can close PRs if no author response in a month.
@14NGiestas <https://github.com/14ngiestas>

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/minetest/minetest/pull/5124#issuecomment-348857057>,
or mute
the thread

**Comment by paramat** [comment]
Yes fine, add a comment in this thread if or when you want it re-opened.

**Pull request closed without merging**